### PR TITLE
Implement Homebrew as the system deployer for macOS.

### PR DESCRIPTION
**Pull request by simonjwright**
See Issue #1172.

It's not clear to me how we could deal with MacPorts as an alternate deployer, if anyone wants to do the work (won't be me).

I'm not sure about  `System_Package_For_Tool` in `src/alire/alire-utils-tools.adb`; I've treated macOS like all the other distributions so as to get a build, but I don't understand how/when this is used.

### CODE REVIEW

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (Fabien-Chouteau) [inline src/alire/alire-origins-deployers-system-homebrew.adb]: Across the project we use `Alire.OS_Lib.Getenv` to get environment variable . We should probably use `Ada.Environment_Variables.Value` but for consistency let's use keep `Alire.OS_Lib.Getenv` everywhere.

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (Fabien-Chouteau) [inline src/alire/alire-origins-deployers-system-homebrew.adb]: Dead code?

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (Fabien-Chouteau) [inline src/alire/alire-origins-deployers-system-homebrew.adb]: Is the `HOMEBREW_PREFIX` env var something common for Homebrew users? Or a requirement only for Alire users?

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (Fabien-Chouteau) [inline src/alire/alire-origins-deployers-system-homebrew.adb]: Detection of the "distrib" should be done in the platform specific code. There's no need to do it here as far as I know.

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (Fabien-Chouteau) [inline src/alire/alire-origins-deployers-system-homebrew.adb]: The `Semantic_Versioning` library is already providing this kind of conversion. Is there a reason not to use it?

File: src/alire/alire-platforms.ads
Comment (Fabien-Chouteau) [inline src/alire/alire-platforms.ads]: This entry should be `Homebew` and not `MacOS`, for the same reasons that we have `Msys2` and not `Windows`:
 - You can use Alire on MacOS without Homebew
 - You can have other system package managers on MacOS (Macports)

File: src/alire/os_macos/alire-platforms-current__macos.adb
Comment (Fabien-Chouteau) [inline src/alire/os_macos/alire-platforms-current__macos.adb]: This is where you detect the `HOMEBREW_PREFIX` env variable and return `Platforms.Homebew` if present, `Platforms.Distro_Unknown` otherwise.

File: testsuite/drivers/helpers.py
Comment (Fabien-Chouteau) [inline testsuite/drivers/helpers.py:96]: This will have to be changed according to the comments about MacOS vs Homebew distrib.

**Review by Fabien-Chouteau — CHANGES_REQUESTED** [review]
Awesome @simonjwright !

I have a few comments, the main one being about the `MacOS` vs `Homebrew` distrib.

Thanks for your contribution,

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (simonjwright) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: OK (but may become OBE depending on how we resolve the Homebrew/MacPorts issue)

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (simonjwright) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: Dead comment, anyway! OK.

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (simonjwright) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: No reason. The existing code will (try to) detect format errors, is it OK to let `Semantic_Versioning` raise an exception in this case? would said exception need to be handled? (it doesn't seem to be handled in other uses).

My code would report libtcl's version as 8.6.12, Semantic_Versioning reports 8.6.12+_1 - would that be a problem? (building my test code with this works fine).

Comment (Fabien-Chouteau) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: @mosteo I think we need your input here.

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (simonjwright) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: OK

File: src/alire/alire-platforms.ads
Comment (simonjwright) [reply src/alire/alire-platforms.ads]: OK

File: src/alire/os_macos/alire-platforms-current__macos.adb
Comment (simonjwright) [reply src/alire/os_macos/alire-platforms-current__macos.adb]: OK

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (simonjwright) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: The 'normal' way for users to set up to use Homebrew is to do something like
```
eval "$(/where/ever/bin/brew shellenv)"
``` 
in their `.bash_profile` or equivalent, where `brew shellenv` returns (on my aarch64 Mac)
```
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```
Of course they could just set the `PATH` themselves. Which means no `HOMEBREW_PREFIX` etc; but then consider [this](https://github.com/simonjwright/tcladashell/blob/8510f86420d0b33bca68441592efc427a7509e5d/options.gpr#L36) and [this](https://github.com/simonjwright/tcladashell/blob/8510f86420d0b33bca68441592efc427a7509e5d/options.gpr#L63) in my tash crate (to be announced), 

The less-obvious issue is that Homebrew on `x86_64-apple-darwin` installs in `/usr/local` by default, but on `aarch64` it installs in `/opt/homebrew` (so as not to confuse people who have both architectures installed???) -- the `/usr/local` thing was my chief reason for rejecting Homebrew; you can change it, e.g. on my x86_64 Mac I have it in `~/homebrew`, which has the disadvantage that packages have to be rebuilt from source rather than being available in binary.

Comment (simonjwright) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: The bottom line is, we can probably get away without setting up these environment variables for `alr` itself (check whether `brew` is on the `PATH`), but a crate that depends on an external will need them.

Comment (Fabien-Chouteau) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: If `eval "$(/where/ever/bin/brew shellenv)"` is the normal way then using `HOMEBREW_PREFIX` sounds like a perfectly good solution to me :+1:

File: testsuite/drivers/helpers.py
Comment (simonjwright) [reply testsuite/drivers/helpers.py:96]: This is for CI, and the CI runners have Homebrew installed (but not the environment variables! see [here](https://github.com/simonjwright/tcladashell/blob/8510f86420d0b33bca68441592efc427a7509e5d/.github/workflows/ci-macos.yml#L42)).

So can I punt the Macports issue for the moment and just return `'HOMEBREW'`?

Comment (Fabien-Chouteau) [reply testsuite/drivers/helpers.py:96]: I think what we might want to do is have a CI run that use Homebrew (`eval $(brew shellenv)`) and one that doesn't. @mosteo  ?

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (simonjwright) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: Actually, I suspect that this code will only get called at all if the `Distribution` is `Homebrew`? in which case the question is moot (but I would have to change the implementation of `Already_Installed `to use the `Installed_Version` found by `Get_Info`, because my lazy implementation relies on knowing `Homebrew_Prefix`).

And the tests for `Homebrew_Present` would be redundant.

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (mosteo) [reply src/alire/alire-origins-deployers-system-homebrew.adb]: The extra build part should not be a problem. We have to be careful if distributions use `-` in a sense different than semantic versioning pre-releases, though, to avoid mis-reporting versions.

These detectors shouldn't raise; in the worst case we prefer the external to be reported as undetected. But then, relaxed semantic versioning will accept anything that starts with a number as a version.

File: testsuite/drivers/helpers.py
Comment (mosteo) [reply testsuite/drivers/helpers.py:96]: Sounds right to me, just as we test with no package manager on linux and with the supported ones too.

File: src/alire/alire-platforms.ads
Comment (Joebeazelman) [reply src/alire/alire-platforms.ads]: Homebrew also supports Windows 10 subsystem and Linux. You might need to create a package matrix or perhaps another environment variable for the parent package manager, i.e. how was Alire installed regardless of platform. This might be a consideration if, for instance, some operation might need the information.

Comment (simonjwright) [reply src/alire/alire-platforms.ads]: > Homebrew also supports Windows 10 subsystem and Linux. You might need to create a package matrix or perhaps another environment variable for the parent package manager, i.e. how was Alire installed regardless of platform. This might be a consideration if, for instance, some operation might need the information.

When you say "You might need to create ..." I hope you don't mean me? This sounds way outside my comfort zone

File: testsuite/drivers/helpers.py
Comment (simonjwright) [reply testsuite/drivers/helpers.py:96]: I could certainly set this up.

I did think about checking whether `brew` is on the `PATH`, but users might want to use a different package manager, & if they had both this could lead to confusion; checking for `HOMEBREW_PREFIX` would mean they were OK with it.

Would have to change `testsuite/drivers/helpers.py` to make the same choice: e.g. like [this](https://github.com/simonjwright/alire/commit/50f4fe9ed5cd6aa037ea9a6558a984bde9bb2b46) and [this](https://github.com/simonjwright/alire/commit/913e5573e3d47557d2e750fe2e6e71f2202a3eec), successful run [here](https://github.com/simonjwright/alire/actions/runs/3077781843)

File: src/alire/alire-platforms.ads
Comment (mosteo) [reply src/alire/alire-platforms.ads]: > When you say "You might need to create ..." I hope you don't mean me? This sounds way outside my comfort zone

I guess it was said as hypothetical future feature. In any case, Alire currently only supports one package manager per distro. But, certainly, one can use combos [os x distro] in the externals, although the need has not yet arisen.

Which makes me think that, as we have `macOS` as the OS, maybe it's more appropriate to use `homebrew` as the distro just like `msys2` is used on Windows. This would enable using other package manager on macOS if someone wants to come up with it.

Comment (simonjwright) [reply src/alire/alire-platforms.ads]: Yes, that’s how I’m approaching it. I’m tidying up the update to this PR.

**Comment by mosteo** [comment]
Looks good to me, outstanding effort, Simon. Since this PR has had quite some discussion, I'm going to wait for @Fabien-Chouteau feedback before merging.

File: src/alire/alire-origins-deployers-system-homebrew.adb
Comment (Fabien-Chouteau) [inline src/alire/alire-origins-deployers-system-homebrew.adb]: All warnings off for the entire file? You can target a specific warning, but disabling all of them is not ok.

**Review by Fabien-Chouteau — CHANGES_REQUESTED** [review]

**Review by Fabien-Chouteau — APPROVED** [review]

**Comment by Fabien-Chouteau** [comment]
Thanks @simonjwright, can you fix the conflicts with the OpenSUSE support that was just added? Sorry for that

**Comment by simonjwright** [comment]
I did fix the conflicts, and hoped to see CI runs; maybe the reason why they didn’t happen is that Github thinks there are still conflicts?

**Comment by Fabien-Chouteau** [comment]
GitHub says there's still one conflic:
```
            System.Zypper.Deployer'(Deployers.Deployer'(Base => From)
<<<<<<< homebrew-merged
                                    with others => <>),
         when Platforms.Homebrew =>
            System.Homebrew.Deployer'(Deployers.Deployer'(Base => From)
                                      with others => <>));
=======
                                    with others => <>));
>>>>>>> master
```

**Comment by simonjwright** [comment]
Actually, Github says there are 3 conflicts, but they are in fact the *changes* that I want to make. I don’t see how I can fix this ...

Oh. Perhaps the problem is that I cherry-picked the openSuse commit from master? ... but that introduced conflicts which I fixed, and I don’t see any other trace of it than the commit (cb3ca3b) author ...

Oh². I can make it happen here. Baffled for the time being

**Comment by mosteo** [comment]
Let me take a stab at this

**Comment by mosteo** [comment]
Sorry for stepping in in your fixing, Simon. Merged now!

**Comment by simonjwright** [comment]
> Sorry for stepping in in your fixing, Simon. Merged now!

Brilliant! Thanks ..

**Pull request merged**
